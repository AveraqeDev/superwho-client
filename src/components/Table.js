import React, { useContext } from 'react';
import { ApiContext } from '../contexts/ApiContext';

import '../styles/Table.css';

const TableControls = React.memo(function TableConrtols() {
  return (
    <div className="TableControls">
      <p>Showing 10 of 113</p>
      <div className="PageButtons">
        <button className="TableControls__button prev">
          {'< Prev'}
        </button>
        <button className="TableControls__button next">
          {'Next >'}
        </button>
      </div>
    </div>
  );
});

const Table = () => {
  const { results } = useContext(ApiContext);

  console.log(results);

  return (
    <div className="Table">
      <TableControls />
      <h1>Data Table goes here</h1>
      {results}
    </div>
  )
};

export default Table;