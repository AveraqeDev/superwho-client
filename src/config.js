import React from 'react';
import Badge from './components/Badge';
import FavoriteButton from './components/FavoriteButton';

export default {
  API_URL: process.env.REACT_APP_API_URL,
  TOKEN_KEY: process.env.REACT_APP_TOKEN_KEY,
  COLUMNS: [
    {
      Header: 'Name',
      accessor: 'name',
      id: 'name',
      defaultCanSort: true
    },
    {
      Header: 'Alignment',
      accessor: 'biography.alignment',
      id: 'alignment',
      defaultCanSort: true
    },
    {
      Header: 'Race',
      accessor: 'appearance.race',
      id: 'race',
      defaultCanSort: true
    },
    {
      Header: 'Gender',
      accessor: 'appearance.gender',
      id: 'gender',
      defaultCanSort: true
    },
    {
      Header: 'Height',
      accessor: 'appearance.height',
      Cell: ({ cell: { value } }) => <Badge values={value} />,
      disableSortBy: true
    },
    {
      Header: 'Weight',
      accessor: 'appearance.weight',
      Cell: ({ cell: { value } }) => <Badge values={value} />,
      disableSortBy: true
    },
    {
      Header: '',
      id: 'actions',
      accessor: 'id',
      Cell: ({ cell: { value } }) => <FavoriteButton heroId={value} />,
      disableSortBy: true
    }
  ]
};