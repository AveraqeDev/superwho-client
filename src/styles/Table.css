.Table {
  display: block;
  max-width: 100%;
  max-height: calc(100vh - 190px);
  overflow-x: scroll;
  overflow-y: scroll;
}

.Table__table {
  text-align: left;
  width: 100%;
  padding: 0 2rem;
}

.Table__th {
  padding: 1rem .2rem;
  color: #8A8888;
  font-weight: 900;
  margin: 0;
  position: sticky;
  top: 0px;
  background: #ffffff;
  border-bottom: 1px solid #C6C7C9;
  z-index: 0;
}

.Table__td {
  padding: .6rem .2rem;
  color: #ADADAD;
  font-weight: 500;
  margin: 0;
  border-bottom: 1px solid #C6C7C9;
}

.TableControls {
  display: flex;
  justify-content: space-between;
  align-items: center;
  width: 100%;
  padding: .2rem 2rem;
  background: #F4F5F7;
  box-shadow: inset 0px 4px 4px rgba(0, 0, 0, 0.25);
}

.TableControls > p {
  color: #C6C7C9;
  font-size: 16px;
  font-weight: bold;
}

.TableControls__button {
  border: none;
  border-radius: 0;
  background: #FFFFFF;
  box-shadow: 0px 4px 4px rgba(0, 0, 0, 0.25);
  padding: 5px 10px;
  color: #C6C7C9;
  font-size: 16px;
  font-weight: bold;
  cursor: pointer;
}

.TableControls__button:hover {
  opacity: 0.8;
}

.TableControls__button:disabled {
  cursor: not-allowed;
  opacity: 0.5;
}

.prev {
  border-top-left-radius: 5px;
  border-bottom-left-radius: 5px;
}

.next {
  border-top-right-radius: 5px;
  border-bottom-right-radius: 5px;
}

@media screen and (max-width: 760px) {
  .Table__table {
    padding: 0 .5rem;
  }

  .Table__th, .Table__td {
    text-align: center;
  }
}

@media screen and (max-width: 350px) {
  .TableControls {
    flex-direction: column;
  }

  .TableControls > p {
    margin-top: .3rem;
  }

  .TableControls__button {
    margin-bottom: .5rem;
  }
}