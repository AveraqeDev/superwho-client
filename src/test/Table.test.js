import React from 'react';
import ReactDOM from 'react-dom';
import renderer from 'react-test-renderer';
import config from '../config';
import Table, { TableControls } from '../components/Table';

describe('Table', () => {
  describe('<TableControls />', () => {
    const testData = {
      pageCount: 3,
      page: 2,
      canPrevious: true,
      onPrev: () => {},
      canNext: true,
      onNext: () => {}
    };
    // Smoke test
    it('renders without crashing', () => {
      const div = document.createElement('div');
      ReactDOM.render(
        <TableControls 
          pageCount={testData.pageCount}
          page={testData.page}
          canPrevious={testData.canPrevious}
          onPrev={testData.onPrev}
          canNext={testData.canNext}
          onNext={testData.onNext}
        />, div);
      ReactDOM.unmountComponentAtNode(div);
    });
  
    //Snapshot test
    it('renders the UI as expected', () => {
      const tree = renderer
        .create(
          <TableControls 
            pageCount={testData.pageCount}
            page={testData.page}
            canPrevious={testData.canPrevious}
            onPrev={testData.onPrev}
            canNext={testData.canNext}
            onNext={testData.onNext}
          />
        )
        .toJSON();
      expect(tree).toMatchSnapshot();
    })
  });

  describe('<Table />', () => {
    const testData = {
      columns: config.COLUMNS,
      data: [
        {
          id: 1,
          name: 'A-Bomb',
          biography: {
            alignment: 'good'
          },
          appearance: {
            race: 'Human',
            gender: 'Male',
            height: [
              "6'8",
              '203 cm'
            ],
            weight: [
              '980 lb',
              '441 kg'
            ]
          }
        },
        {
          id: 2,
          name: 'Abe Sapien',
          biography: {
            alignment: 'good'
          },
          appearance: {
            race: 'Icthyo Sapien',
            gender: 'Male',
            height: [
              "6'3",
              '191 cm'
            ],
            weight: [
              '145 lb',
              '65 kg'
            ]
          }
        }
      ]
    };
    // Smoke test
    it('renders without crashing', () => {
      const div = document.createElement('div');
      ReactDOM.render(
        <Table columns={testData.columns} data={testData.data} />, div);
      ReactDOM.unmountComponentAtNode(div);
    });
  
    //Snapshot test
    it('renders the UI as expected', () => {
      const tree = renderer
        .create(
          <Table columns={testData.columns} data={testData.data} />
        )
        .toJSON();
      expect(tree).toMatchSnapshot();
    })
  });
});