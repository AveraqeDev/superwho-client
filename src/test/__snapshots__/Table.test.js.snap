// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Table <Table /> renders the UI as expected 1`] = `
Array [
  <div
    className="TableControls"
  >
    <p>
      Page 
      1
       of 
      1
    </p>
    <div
      className="PageButtons"
    >
      <button
        className="TableControls__button prev"
        disabled={true}
        onClick={[Function]}
      >
        &lt; Prev
      </button>
      <button
        className="TableControls__button next"
        disabled={true}
        onClick={[Function]}
      >
        Next &gt;
      </button>
    </div>
  </div>,
  "",
  <div
    className="Table"
  >
    <table
      className="Table__table"
      role="table"
    >
      <thead
        className="Table__thead"
      >
        <tr
          className="Table__tr"
          role="row"
        >
          <th
            className="Table__th"
            colSpan={1}
            onClick={[Function]}
            role="columnheader"
            style={
              Object {
                "cursor": "pointer",
              }
            }
            title="Toggle SortBy"
          >
            Name
            <span
              className="SortArrow"
            >
               ▲
            </span>
          </th>
          <th
            className="Table__th"
            colSpan={1}
            onClick={[Function]}
            role="columnheader"
            style={
              Object {
                "cursor": "pointer",
              }
            }
            title="Toggle SortBy"
          >
            Alignment
            <span
              className="SortArrow"
            >
              
            </span>
          </th>
          <th
            className="Table__th"
            colSpan={1}
            onClick={[Function]}
            role="columnheader"
            style={
              Object {
                "cursor": "pointer",
              }
            }
            title="Toggle SortBy"
          >
            Race
            <span
              className="SortArrow"
            >
              
            </span>
          </th>
          <th
            className="Table__th"
            colSpan={1}
            onClick={[Function]}
            role="columnheader"
            style={
              Object {
                "cursor": "pointer",
              }
            }
            title="Toggle SortBy"
          >
            Gender
            <span
              className="SortArrow"
            >
              
            </span>
          </th>
          <th
            className="Table__th"
            colSpan={1}
            role="columnheader"
            style={
              Object {
                "cursor": undefined,
              }
            }
          >
            Height
            <span
              className="SortArrow"
            >
              
            </span>
          </th>
          <th
            className="Table__th"
            colSpan={1}
            role="columnheader"
            style={
              Object {
                "cursor": undefined,
              }
            }
          >
            Weight
            <span
              className="SortArrow"
            >
              
            </span>
          </th>
          <th
            className="Table__th"
            colSpan={1}
            role="columnheader"
            style={
              Object {
                "cursor": undefined,
              }
            }
          >
            
            <span
              className="SortArrow"
            >
              
            </span>
          </th>
        </tr>
      </thead>
      <tbody
        className="Table__tbody"
        role="rowgroup"
      >
        <tr
          className="Table__tr"
          role="row"
        >
          <td
            className="Table__td"
            role="cell"
          >
            A-Bomb
          </td>
          <td
            className="Table__td"
            role="cell"
          >
            good
          </td>
          <td
            className="Table__td"
            role="cell"
          >
            Human
          </td>
          <td
            className="Table__td"
            role="cell"
          >
            Male
          </td>
          <td
            className="Table__td"
            role="cell"
          >
            <span
              className="badge"
            >
              6'8
            </span>
            <span
              className="badge"
            >
              203 cm
            </span>
          </td>
          <td
            className="Table__td"
            role="cell"
          >
            <span
              className="badge"
            >
              980 lb
            </span>
            <span
              className="badge"
            >
              441 kg
            </span>
          </td>
          <td
            className="Table__td"
            role="cell"
          >
            <span
              className="favme"
              onAnimationEnd={[Function]}
              onClick={[Function]}
            >
              ♥
            </span>
          </td>
        </tr>
        <tr
          className="Table__tr"
          role="row"
        >
          <td
            className="Table__td"
            role="cell"
          >
            Abe Sapien
          </td>
          <td
            className="Table__td"
            role="cell"
          >
            good
          </td>
          <td
            className="Table__td"
            role="cell"
          >
            Icthyo Sapien
          </td>
          <td
            className="Table__td"
            role="cell"
          >
            Male
          </td>
          <td
            className="Table__td"
            role="cell"
          >
            <span
              className="badge"
            >
              6'3
            </span>
            <span
              className="badge"
            >
              191 cm
            </span>
          </td>
          <td
            className="Table__td"
            role="cell"
          >
            <span
              className="badge"
            >
              145 lb
            </span>
            <span
              className="badge"
            >
              65 kg
            </span>
          </td>
          <td
            className="Table__td"
            role="cell"
          >
            <span
              className="favme"
              onAnimationEnd={[Function]}
              onClick={[Function]}
            >
              ♥
            </span>
          </td>
        </tr>
      </tbody>
    </table>
  </div>,
]
`;

exports[`Table <TableControls /> renders the UI as expected 1`] = `
<div
  className="TableControls"
>
  <p>
    Page 
    3
     of 
    3
  </p>
  <div
    className="PageButtons"
  >
    <button
      className="TableControls__button prev"
      disabled={false}
      onClick={[Function]}
    >
      &lt; Prev
    </button>
    <button
      className="TableControls__button next"
      disabled={false}
      onClick={[Function]}
    >
      Next &gt;
    </button>
  </div>
</div>
`;
